Remove estimations where score data is available for osu! difficulty calculations

Finadoggie opened the pull request:
This PR replaces estimations with score data where applicable for Lazer scores. More specifically, it…
1. Modifies estimations for slider breaks 
2. Removes estimations for slider end drops

For removing slider break estimations: 
- `HitResult.Miss` includes both misses and slider head misses, accounting for two types of combo breaks
- `HitResult.LargeTickMiss` includes dropped reverse arrows and dropped slider ticks
- For Lazer scores, large tick misses are deliberately capped to avoid large amounts of them nuking a score, as buzz sliders and high slider tick rates can easily reward players with numerous amounts of large tick misses
- Scores with no large tick misses will have no estimated misses, removing an issue where lazer scores were unfairly punished for theoretical slider breaks that were demonstrably not present

For removing slider end drop estimations:
- Slider end drops and large tick misses are used in place of estimated slider end drops
- Calculations which use estimated slider end drops are otherwise left untouched

Notes: 
- Classic mod is not being considered as ppy has previously said that pp for classic mod should match stable

=== original post ===
(contains outdated info)

On non-CL lazer scores, extra slider data means certain estimations are no longer required.

For effective miss count, HitResults.Miss + HitResults.LargeTickMiss is used.

For sliderends dropped, HitResults.SmallTickMiss is used.

To my understanding, LargeTickMiss includes slider ticks and reverse arrows, the two parts that can break your combo, while SmallTickMiss is for sliderends. Please correct me if I am wrong.

Finadoggie committed:
Make length bonus account for sliders, use proper misscount for classic

Finadoggie committed:
Use actual sliderends dropped instead of estimating

Score data for non-CL scores includes sliderends dropped, meaning no need to estimate.

CL scores are still estimated.

Finadoggie committed:
Revert "Make length bonus account for sliders, use proper misscount for classic"

This reverts commit 941c0487a454fad1447812f2cd12fad26f1cd28c.

Finadoggie committed:
Use miss count for effective miss count

No need to estimate misses for non-CL scores.

Finadoggie committed:
Merge pull request #1 from Finadoggie/miss-count-fix

Miss count fix

Finadoggie committed:
Merge branch 'estimation-removal' into dropped-tail-fix

Finadoggie committed:
Merge pull request #2 from Finadoggie/dropped-tail-fix

Use actual sliderends dropped instead of estimating

Finadoggie committed:
Update OsuPerformanceCalculator.cs

Finadoggie committed:
Add slider ticks and reverse arrows to effective misscount

Very much open to discussion on if these should be weighed differently

bdach commented:
I am not sure the "available" score data _can_ be used. We've previously expressed worry that this was going to spiral into two separate implementations of pp for classic and non-classic scores.

Finadoggie commented:
My bad, I was not aware of these conversations. Could you drop a link to them so I can read through them?

bdach commented:
https://github.com/ppy/osu/pull/21938 would be a good starting point I guess

Natelytle commented:
Stan's decision of splitting the calculator into 2 parts for classic and non-classic is a direction I personally dislike, and the way Fina is doing it here is the way I would approach it myself. I don't see any harm in using the new metrics, and PP spiraling into 2 calculators should only happen if per note judgements ever become available, and prove to give much better results than without them. These new metrics are easily estimable, and the existing estimations can just be made harsher if they prove to give an advantage to stable scores.

Flamiii commented:
The implications of a change like this are massive. If I set scores with sliderbreaks, I will only ever be punished more on lazer than I would be if I played on stable. I play on lazer as my main client, but a change like this on its own would make me consider moving back to stable because I don't want to play with something like this that is just a clear disadvantage. You could try tuning the estimations for stable scores to be more harsh, but I doubt people would be happy with that and I'm honestly not sure if that's even possible. The complexity to balance this kind of change is so high that I don't think it's worth the trouble to begin with.

Finadoggie commented:
@Flamiii you’re already at an objective disadvantage if you play on lazer. This arguably helps you since it means miss estimations can no longer assume you may have sliderbroke where you didn’t. Any non-cl 1 miss plays where the break was in the middle are buffed. Same for plays on hard slider maps where you broke but didn’t drop any sliderends.

Flamiii commented:
If by "objective disadvantage" you mean slideracc, I play on lazer because I'm certain that will be taken into account in the future (#27063). The scenario I'm thinking of which happens quite a bit is the one where a player misses near the end of a map but also sliderbreaks 2 times afterwards. In stable, this play is counted as a 1 miss and the sliderbreaks only negatively impact your accuracy. With these changes in lazer, this play would be treated as if it was a 3 miss. As far as I know, there really isn't a way to adjust stable's estimation to make this scenario balanced because the relevant information just isn't there. There are some situations where you're punished a bit less, but the scenario I mentioned is extremely common and you would be punished way more for playing on lazer with these changes.

Rekunan commented:
> there really isn't a way to adjust stable's estimation

There actually is a way, if you take a look at [the function in question](https://github.com/ppy/osu/blob/77660e57ea3d675b92d4c6c0c5eb8e68562d9135/osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs#L249), we could, for example, lower fullComboThreshold to make it assume there were more misses than before, and if you look at the changes in this pr, this function is only used if CL with slideracc is on.

> misaiming a buzz slider with 5 repeats ... would basically kill any score

This is a valid point, however, I can assure you that future work will be done to account for this.

Flamiii commented:
> There actually is a way, if you take a look at [the function in question](https://github.com/ppy/osu/blob/77660e57ea3d675b92d4c6c0c5eb8e68562d9135/osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs#L249), we could, for example, lower fullComboThreshold to make it assume there were more misses than before, and if you look at the changes in this pr, this function is only used if CL with slideracc is on.

I see. As long as this ends up being balanced properly, I have no issues with it.

Finadoggie commented:
> If by "objective disadvantage" you mean slideracc, I play on lazer because I'm certain that will be taken into account in the future (#27063). The scenario I'm thinking of which happens quite a bit is the one where a player misses near the end of a map but also sliderbreaks 2 times afterwards. In stable, this play is counted as a 1 miss and the sliderbreaks only negatively impact your accuracy. With these changes in lazer, this play would be treated as if it was a 3 miss. As far as I know, there really isn't a way to adjust stable's estimation to make this scenario balanced because the relevant information just isn't there. There are some situations where you're punished a bit less, but the scenario I mentioned is extremely common and you would be punished way more for playing on lazer with these changes.
> 
> Edit: It's also worth noting that using HitResults.LargeTickMiss to increase the effective miss count means that missing buzz sliders or any fast sliders with many sliderticks will be incredibly punishing. For example, misaiming a buzz slider with 5 repeats and missing it entirely would increase the effective miss count by 6, which would basically kill any score.

Finadoggie commented:
shoot I hit the wrong button

Finadoggie commented:
> The scenario I'm thinking of which happens quite a bit is the one where a player misses near the end of a map but also sliderbreaks 2 times afterwards.
Ok but again this is already a thing. Sliderbreaks by missing the slider head are already counted as misses in lazer, so this scenario can already happen without any pp changes.
> It's also worth noting that using HitResults.LargeTickMiss to increase the effective miss count means that missing buzz sliders or any fast sliders with many sliderticks will be incredibly punishing. For example, misaiming a buzz slider with 5 repeats and missing it entirely would increase the effective miss count by 6, which would basically kill any score.
I know, I'm more apprehensive about that part and wanted to spark discussion for that specifically.

Finadoggie committed:
Merge branch 'master' into estimation-removal

Finadoggie committed:
Use sliderend data for all non-legacy scores

As per suggestion by givikap, I was not aware that non-legacy cl scores stored this data

Finadoggie committed:
Merge branch 'estimation-removal' of https://github.com/Finadoggie/osu into estimation-removal

Finadoggie committed:
Cap Buzz Slider Related Misses

After letting the comments @Flamiii left brew for a while, I realized they were very much right about the buzz slider thing. As such, I've implemented a quick and dirty untested fix that will hopefully have zero unintended side-effects :)

I don't see this as a permanent or final solution yet. There's definitely some potential issues/inaccuracies that could arise with maps like Notch Hell or IOException's Black Rover, but afaik this implementation would not cause any issues that stable doesn't already have.

Finadoggie committed:
removed large tick misses from effectivemisscount

Givikap120 commented:
https://github.com/ppy/osu/discussions/27831

Finadoggie commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -39,8 +41,14 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+            if (!score.Mods.Any(h => h is OsuModClassic cl && cl.NoSliderHeadAccuracy.Value))
+                effectiveMissCount = countMiss; 
+            else
+                effectiveMissCount = calculateEffectiveMissCount(osuAttributes);
```
Any reason why…?

Givikap120 replied:
you already used amount missed sliderends in lazer CL scores
but now you're trying to use estimator (that's almost always outputting lower value than normal misscount)
this doesn't makes any sense unless your objective is made lazer CL objectively the best way to farm, so slider-acc players will suffer

Finadoggie commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -122,8 +130,14 @@ private double computeAimValue(ScoreInfo score, OsuDifficultyAttributes attribut
+                if (score.IsLegacyScore)
+                    estimateSliderEndsDropped = Math.Clamp(Math.Min(countOk + countMeh + countMiss, attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders);
+                else
+                    estimateSliderEndsDropped = countSliderEndsDropped;
```
1. What about this is incorrect?

2. I’m hesitant to do that since I’d imagine values using `estimateSliderEndsDropped` are balanced around sliderend drops, not sliderend drops + other stuff

Givikap120 replied:
what if you hit all sliderends but missed all reverses?
you will still get full slideraim value?

Givikap120 replied:
> 1. What about this is incorrect?
> 2. I’m hesitant to do that since I’d imagine values using `estimateSliderEndsDropped` are balanced around sliderend drops, not sliderend drops + other stuff

also, it's not balanced around sliderend drops, because in case where you not dropped sliderend but sliderbroke - you almost instantly loosing almost all slider pp

Givikap120 replied:
also change it to `Math.Min(countSliderEndsDropped + countLargeTickMiss, estimateDifficultSliders);`
because otherwise you can get negative pp by missing to many sliderends

Finadoggie commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -39,8 +41,14 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+            countLargeTickMiss = score.Statistics.GetValueOrDefault(HitResult.LargeTickMiss);
+            countSliderEndsDropped = score.Statistics.GetValueOrDefault(HitResult.SmallTickMiss);
+            if (!score.Mods.Any(h => h is OsuModClassic cl && cl.NoSliderHeadAccuracy.Value))
+                effectiveMissCount = countMiss; 
```
If you can back up that claim of "almost always", I’ll consider this. I know for a fact that I have some scores that would’ve benefitted from this change had they been set in lazer, and there’s definitely some famous scores that would have benefitted as well (e.g. mou ii kai)

Givikap120 replied:
literally like 99% of my nonFC scores have higher misscount in lazer compared to stable
this should be the most obvious thing ever, considering that in stable you're just ignoring sliderbreaks in misscount almost completely

Finadoggie replied:
I would prefer if you could keep snide remarks such as "this should be the most obvious thing ever" out of my PRs. Insulting people will only ever detract from your suggestions. 

That said, I see what you’re saying, but I’m worried that lowering the value of a miss based on slider count will be abused by pp mappers in some form.

Givikap120 replied:
okay, do the opposite thing - increase the miss penalty for not slideracc scores depending on how many sliders here

Givikap120 replied:
> I would prefer if you could keep snide remarks such as "this should be the most obvious thing ever" out of my PRs. Insulting people will only ever detract from your suggestions.
> 
> That said, I see what you’re saying, but I’m worried that lowering the value of a miss based on slider count will be abused by pp mappers in some form.

also, "the most obvious thing ever" was said from the fact that you have one example (mou ii kai) of effectiveMissCount being more harsh, when ignoring the fact that there are hundreds of other examples of it being more lenient, what is kinda weird cherry-picking from your side

Finadoggie replied:
> okay, i agree with this, then do the opposite thing - increase the miss penalty for not slideracc scores depending on how many sliders here
to compensate the fact that all scores with misses are overnerfed - you can decrease miss penalty

This still falls into the same trap of changing the value of a miss based on slider count. I don't want to do that. Plus, I don't feel like nerfing every stable score with misses is reasonable.

> "the most obvious thing ever" was said because of the fact that you have one example (mou ii kai) of effectiveMissCount being more harsh, while not mentioning the fact that there are hundreds of other examples of it being more lenient, what is kinda weird cherry-picking from your side

The Mou Ii Kai example was just one I pulled off the top of my head, I have plenty of my own scores for which this same idea would apply. That said, I don't think fighting anecdotes with anecdotes is particularly productive, especially when we can actually determine decently well which scores would be nerfed, which scores would be buffed, and which scores would stay the same

Scores which would be buffed
- Scores with just one miss/sb in the middle
- Scores with relatively evenly distributed misses, e.g. two misses/sb scores where one miss was ~1/3 into the map and the other was ~2/3 into the map

Scores which would stay the same
- Chokes with just one miss/sb near the end of the map
- Scores with several misses, especially with high combo, but no sbs 

Scores which would be nerfed 
- Scores with numerous sliderbreaks, more than `effectiveMissCount` can account for
- Scores with high combo, and some or many sliderbreaks
- Both cases can (and often will) also include misses

Now, I could bring up several examples of scores which would be buffed or stay the same, but I don't think the examples are what matter. None of the parameters for those categories are particularly hard to meet, and if any of the proposals that remove combo scaling are pushed, we will most definitely see more scores that fall into the buffed category. 

Regardless, many of the scores that fall into the nerfed category wouldn't be worth much regardless as that's just the nature of high misscount scores. If there's ever a reality in which this isn't true, maybe this could be revisited, but for now, I'd consider those scores to be largely negligible. 

If you know that you're going to set a score with numerous sliderbreaks, CL will be a better choice no matter what. There simply isn't a way to avoid this without being unfair to stable scores without numerous sliderbreaks. However, I think that's fine. I've outlined the scenarios in which non-CL is the same or better, and for people going for FCs or low misscounts (which are really the only scores that give pp), I feel that the insurance of knowing you won't be screwed over by sliderbreak estimations is enough to justify using it instead of CL.

Givikap120 replied:
Okay, "several examples of scores"? Let's see mrekk plays (non FCs that have video or replay), and how many of them will be nerfed if they were set in lazer:
- Save me: nerfed
- Space battle: same (would be heavily nerfed if he broke slightly later)
- Owari: nerfed
- Anoyo iki: nerfed
- Yakata mawari: same
- Ghost rule: same
- Brazil: same
- Save me: nerfed
- TFaF: same
- Fur Elisa: same
- Rage of Dust: nerfed
Okay, i will stop here. I can say that there are NOT A SINGLE play in mrekk top100 that is buffed by the using true number of misses. 
Yea, this is because of the combo scaling. What about mrekk high pp scores in combo scaling removal?
- Owari 3 mod: nerfed
- Uta: same
- Yakata mawari: buffed
- Yomi Yori: nerfed
- Inai Sekai (2 miss): nerfed
- Packet Hero: nerfed
Even in your best case - scores with misses in the middle, are overnerfed by your lazer pp
Therefore your claim about "plenty of my own scores" is absolutely baseless and not based on reality (out of 17 scores - 9 are nerfed, 7 stayed the same and 1 is buffed)

Givikap120 replied:
You know the situation with the slideracc? The majority of community not like this at all, and by making lazer pp that 
 harsh you're making the situation where like 95% of the people will play either stable or lazer with CL.

Givikap120 replied:
>If you know that you're going to set a score with numerous sliderbreaks, CL will be a better choice no matter what. There simply isn't a way to avoid this without being unfair to stable scores without numerous sliderbreaks. However, I think that's fine. I've outlined the scenarios in which non-CL is the same or better, and for people going for FCs or low misscounts (which are really the only scores that give pp), I feel that the insurance of knowing you won't be screwed over by sliderbreak estimations is enough to justify using it instead of CL.

You're making it "CL is a better choice no matter what, except super rare edgecase", this is very bad when mod that is considered to be mod for boomers - is objectively the best way to farm for everyone

Givikap120 replied:
It's lazer non-CL what should be slightly superior if anything (if equality is not possible)

Finadoggie replied:
I’m just gonna resolve this since I’m not open to change on this matter for the time being. This PR is just meant to be an initial step, not a final balancing pass. We can revisit this later.

Givikap120 commented:
Actually, new score data can allow using 3 types of aim instead of 2:
- Only notes
- Notes + LargeSliderTicks (reverses and sliderpoints)
- Notes + LargeSliderTicks + SliderEnds
But this is out of scope of this PR

Givikap120 commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -39,8 +41,14 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
             countMiss = score.Statistics.GetValueOrDefault(HitResult.Miss);
-            effectiveMissCount = calculateEffectiveMissCount(osuAttributes);
+            countLargeTickMiss = score.Statistics.GetValueOrDefault(HitResult.LargeTickMiss);
+            countSliderEndsDropped = score.Statistics.GetValueOrDefault(HitResult.SmallTickMiss);
```
This thing works
```
if (!score.Mods.Any(h => h is OsuModClassic cl && cl.NoSliderHeadAccuracy.Value))
{
    countSliderEndsDropped = osuAttributes.SliderCount - score.Statistics.GetValueOrDefault(HitResult.SliderTailHit);
}

cdwcgt commented:
I think we can add a `bool useSliderHead`
Because we are likely to use this value frequently in future
like taiko's `isConvert`
https://github.com/ppy/osu/blob/15d286e57d8acc9d3967b63857c2cb0bfa283230/osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs#L45

Finadoggie committed:
Add bool useSliderHead

Finadoggie committed:
Fix getting slider head drops

Finadoggie committed:
Clamp estimatedSliderEndsDrop

Finadoggie committed:
Re-add bool useSliderHead

oops

Finadoggie commented:
useSliderHead is used for sliderend stuffs to ensure compliance with this statement from ppy 
![image](https://github.com/ppy/osu/assets/75299710/d0e7aca6-7a8c-4d96-b76a-e5ffd77f56a7)

Finadoggie committed:
fix code formatting

cdwcgt commented:
useClassicSlider may better if it not only affects the slider head?

Finadoggie committed:
Renamed useSliderHead to useClassicSlider (and refactored code accordingly)

Finadoggie committed:
merged givi's accuracy changes

stat acc save me

Finadoggie committed:
Revert "merged givi's accuracy changes"

This reverts commit 1f55c1413bc5fddbae2d9fc932376e189c9c9023.

Finadoggie committed:
Merge branch 'master' into estimation-removal

stanriders commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -32,14 +36,23 @@ public OsuPerformanceCalculator()
         protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo score, DifficultyAttributes attributes)
         {
             var osuAttributes = (OsuDifficultyAttributes)attributes;
+            useClassicSlider = score.Mods.Any(h => h is OsuModClassic cl && cl.NoSliderHeadAccuracy.Value);
```
`usingClassicSliderMechanics` sounds better to me

Finadoggie replied:
resolved by #27063 creating variable usingClassicSliderAccuracy

stanriders commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -122,8 +135,14 @@ private double computeAimValue(ScoreInfo score, OsuDifficultyAttributes attribut
+                if (useClassicSlider)
+                    estimateSliderEndsDropped = Math.Clamp(Math.Min(countOk + countMeh + countMiss, attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders);
+                else
+                    estimateSliderEndsDropped = Math.Min(countSliderEndsDropped + countLargeTickMiss, estimateDifficultSliders);
```
I'm not sure `min(drops, difficultSliders)` is necessary here since we definitely know how many dropped sliders there are

Finadoggie replied:
I just did it this way to match live since stable scores are clamped in the same way

Givikap120 replied:
> I'm not sure `min(drops, difficultSliders)` is necessary here since we definitely know how many dropped sliders there are. It'd be better to adjust slider nerf curve instead imo

it is necessary because amount of dropped sliderends can be higher than `difficultSliders`, as `difficultSliders` is only 15% of the whole sliders count

Finadoggie replied:
I'm still inclined to keep it as is since I don't want to change the curve and I *really* don't want to open the gates for negative pp values

Finadoggie replied:
on second thought maybe we should change the curve

stanriders replied:
I think keeping it as is is better for now

stanriders commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -32,14 +36,23 @@ public OsuPerformanceCalculator()
+            if (!useClassicSlider)
+                countSliderEndsDropped = osuAttributes.SliderCount - score.Statistics.GetValueOrDefault(HitResult.SliderTailHit);
+
+            if (useClassicSlider)
```
Do stable scores have `NoSliderHeadAccuracy` set?

Finadoggie replied:
idk but the desired effects are achieved when using the perfcalc tools so it seems to be fine?

cdwcgt replied:
If we are worried about this.
we can add `score.IsLegacyScore` to conditions to just in case.

```cs
usingClassicSliderMechanics = score.Mods.Any(h => h is OsuModClassic cl && cl.NoSliderHeadAccuracy.Value) || score.IsLegacyScore
```

Givikap120 replied:
> If we are worried about this. we can add `score.IsLegacyScore` to conditions to just in case.
> 
> ```cs
> usingClassicSliderMechanics = score.Mods.Any(h => h is OsuModClassic cl && cl.NoSliderHeadAccuracy.Value) || score.IsLegacyScore
> ```

no need to worry about this
you can't have stable score without CL unless someone replay-edited it specifically

Finadoggie replied:
just curious, is there any real reason to not have it?

Givikap120 replied:
> just curious, is there any real reason to not have it? (just as an extra safety guard)
the potential reason is when you have some score simulator like Osu-Tools, and you don't wanna see additional `IsLegacy` checkboxes everywhere (you don't have to if everything is done correctly but still)

Finadoggie replied:
resolving because #27063 is merged so I assume it's fine :)

if I'm wrong, just reopen this

stanriders commented:
Also please update the op because I got really confused by what am I even looking at at first

Finadoggie commented:
> Also please update the op because I got really confused by what am I even looking at at first

done, new one should be far less confusing

tsunyoku commented:
@Finadoggie for my own understanding while I try to trample through these changes, do your changes account for CL (on default configuration, which matches stable and doesn't have slider head accuracy) being 1:1 with stable?

Finadoggie commented:
![image](https://github.com/ppy/osu/assets/75299710/ad0191c9-3131-4d29-a91b-1be6b4b82aa6)
dropping this here so it doesn't get lost

Finadoggie committed:
Merge branch 'ppy:master' into estimation-removal

Finadoggie committed:
Merge branch 'ppy:master' into estimation-removal

Finadoggie committed:
Update to use variable usingClassicSliderAccuracy

Finadoggie committed:
Convert estimateSliderEndsDropped assignment into '?:' expression
I would just like to say that I don't know why anyone would ever want this but github told me to do it

stanriders commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -124,8 +134,11 @@ private double computeAimValue(ScoreInfo score, OsuDifficultyAttributes attribut
-                double estimateSliderEndsDropped = Math.Clamp(Math.Min(countOk + countMeh + countMiss, attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders);
-                double sliderNerfFactor = (1 - attributes.SliderFactor) * Math.Pow(1 - estimateSliderEndsDropped / estimateDifficultSliders, 3) + attributes.SliderFactor;
+                double estimateSliderEndsDropped;
+                estimateSliderEndsDropped = usingClassicSliderAccuracy ? Math.Clamp(Math.Min(countOk + countMeh + countMiss, attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders) : Math.Min(countSliderEndsDropped + countLargeTickMiss, estimateDifficultSliders);
```
This is a pain to read, please use normal `if-else`

Finadoggie replied:
I agree but making it not a ternary causes it to fail the code quality test. 
![image](https://github.com/user-attachments/assets/3332eecb-4cfa-4ac0-ade9-a635a3ed5e6d)
Is passing this test not a requirement?

stanriders replied:
Oh, that's kinda stupid. I guess another option would be at least adding some linebreaks instead of having it as one giant line

smoogipoo replied:
It is, and I wouldn't be against relaxing that rule. How do @peppy @bdach feel about relaxing the convert-to-ternary warning on if-statements? I find that its relative use is pretty subjective.

peppy replied:
Can be changed to a hint if you want.

But usually we just use newlines to make this as-legibile-as-if?

```csharp
                estimateSliderEndsDropped = usingClassicSliderAccuracy
                    ? Math.Clamp(Math.Min(countOk + countMeh + countMiss, attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders)
                    : Math.Min(countSliderEndsDropped + countLargeTickMiss, estimateDifficultSliders);
```

Either or, just feels weird to change the rule after being okay for this long.

stanriders commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -124,8 +134,11 @@ private double computeAimValue(ScoreInfo score, OsuDifficultyAttributes attribut
-                double estimateSliderEndsDropped = Math.Clamp(Math.Min(countOk + countMeh + countMiss, attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders);
-                double sliderNerfFactor = (1 - attributes.SliderFactor) * Math.Pow(1 - estimateSliderEndsDropped / estimateDifficultSliders, 3) + attributes.SliderFactor;
+                double estimateSliderEndsDropped;
+                estimateSliderEndsDropped = usingClassicSliderAccuracy ? Math.Clamp(Math.Min(countOk + countMeh + countMiss, attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders) : Math.Min(countSliderEndsDropped + countLargeTickMiss, estimateDifficultSliders);
```
Also please join it with the variable declaration

stanriders commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -124,8 +134,11 @@ private double computeAimValue(ScoreInfo score, OsuDifficultyAttributes attribut
+                estimateSliderEndsDropped = usingClassicSliderAccuracy ? Math.Clamp(Math.Min(countOk + countMeh + countMiss, attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders) : Math.Min(countSliderEndsDropped + countLargeTickMiss, estimateDifficultSliders);
+
+                double sliderNerfFactor = 0;
+                sliderNerfFactor = (1 - attributes.SliderFactor) * Math.Pow(1 - estimateSliderEndsDropped / estimateDifficultSliders, 3) + attributes.SliderFactor;
```
Same here

Finadoggie committed:
Join variable assignments with declarations

Finadoggie committed:
consolidated if statements for getting effectiveMissCount and countSliderEndsDropped

Finadoggie committed:
only assign countLargeTickMiss for slider accuracy scores
helps indicate it should only be used for slider acc scores

Finadoggie committed:
remove score debugging code I accidentally left in

Finadoggie committed:
Revert "remove score debugging code I accidentally left in"

This reverts commit 5192599543a17a9d47a92627d0128051c6904e56.

smoogipoo commented:
!diffcalc

smoogipoo commented:
!diffcalc

smoogipoo commented:
!diffcalc

Finadoggie commented:
is it normal for that diffcalc command to fail four times?

tsunyoku commented:
@smoogipoo ?

stanriders commented:
!diffcalc

smoogipoo commented:
!diffcalc

Finadoggie commented:
recording for safekeeping, will update periodically (ppy pls add a way to view this on the site)

score - [slider tick] [slider end]
vince kaede - 1791/1813 306/311
jan glin insomnia - 60/61 179/182
2oomer state of mind - 161/162 296/306
arakii state of mind - 162/162 284/306

Finadoggie committed:
use LargeTickHit instead of LargeTickMiss
LargeTickMiss appears to not be stored

smoogipoo commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -44,7 +46,16 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
             countMiss = score.Statistics.GetValueOrDefault(HitResult.Miss);
-            effectiveMissCount = calculateEffectiveMissCount(osuAttributes);
+
+            if (usingClassicSliderAccuracy)
+                effectiveMissCount = calculateEffectiveMissCount(osuAttributes);
+            else
+            {
+                countSliderEndsDropped = osuAttributes.SliderCount - score.Statistics.GetValueOrDefault(HitResult.SliderTailHit);
+                // Note: HitResult.LargeTickMiss is not stored within the API or in score dumps. Do not use it!
+                countLargeTickMiss = score.MaximumStatistics.GetValueOrDefault(HitResult.LargeTickHit) - score.Statistics.GetValueOrDefault(HitResult.LargeTickHit);
+                effectiveMissCount = countMiss;
```
I have a few issues here.

## First, this needs a unit test...

It is exceedingly difficult to parse and mentally prove whether any of this is correct. If that means you have to extract this into a `public` method and call it from a unit test, then so be it.

## Secondly, about `LargeTickMiss` storage...

Where is the assertion coming from that LTM is not stored in the API/score dumps coming from? Can you provide specific scores where this is not the case? You'll find that those scores will _also_ not store LTH, so I'm not sure this code is correct at all.

## Thirdly, regarding differences in stored statistics...

Here's a sampling of scores:

stable:
```json
{
  "data": {
    "mods": [
      {
        "acronym": "CL"
      }
    ],
    "statistics": {
      "ok": 25,
      "miss": 6,
      "great": 304
    },
    "maximum_statistics": {
      "great": 335,
      "legacy_combo_increase": 147
    }
  }
}
```

I believe the existing path is fine here - this will go through the estimator and extrapolate via slider count + combo.

lazer:
```json
{
  "data": {
    "statistics": {
      "ok": 228,
      "meh": 96,
      "miss": 531,
      "great": 984,
      "ignore_hit": 527,
      "ignore_miss": 469,
      "large_tick_hit": 498,
      "large_tick_miss": 89,
      "slider_tail_hit": 384
    },
    "maximum_statistics": {
      "great": 1839,
      "ignore_hit": 690,
      "large_tick_hit": 587,
      "slider_tail_hit": 690
    }
  }
}
```

This is where things get shaky for me.
- We can calculate dropped ends exactly via `slider_tail_hit`, and I would highly prefer that over the attributes.
- I don't understand why `large_tick_miss` would be recalculated here. This is why I'm requesting more info with the specific scores that led to the comment.

lazer + CL (classic sliders):
```json
{
  "data": {
    "mods": [
      {
        "acronym": "CL"
      }
    ],
    "statistics": {
      "ok": 65,
      "meh": 4,
      "miss": 4,
      "great": 587,
      "ignore_miss": 10,
      "large_bonus": 46,
      "small_bonus": 95,
      "small_tick_hit": 377,
      "small_tick_miss": 13,
      "large_tick_hit": 412,
      "large_tick_miss": 6
    },
    "maximum_statistics": {
      "great": 660,
      "large_bonus": 56,
      "small_bonus": 95,
      "large_tick_hit": 418,
      "small_tick_hit": 390
    }
  }
}
```

This one is perhaps the worst of the bunch.  
Right now this will go through the estimator, but we can do better because in this case dropped slider ends are represented as `small_tick_miss`.

You can use `score.IsLegacyScore` to tell this score apart from the stable one.

tsunyoku replied:
If I'm remembering correctly, there was a notion that CL and stable scores should match 1:1 in PP so that was the intention with sending all classic scores down the estimation path.

smoogipoo replied:
Okay, that's fair enough.

Finadoggie replied:
![image](https://github.com/user-attachments/assets/f6c13c71-408a-4311-97c3-01178fa6ba92)
Can confirm the notion that lazer CL scores should match stable.

\s for why I switched to using `LargeTickHit`, I was operating mainly off of these messages:
(data.ppy.sh)
![IMG_5770](https://github.com/user-attachments/assets/a561ccc8-5c96-415c-bbdc-590b3b8d7b0d)
(API)
![IMG_5771](https://github.com/user-attachments/assets/e6a71e7e-2982-422b-95a1-a0041cf89dcf)
According to tsunyoku, I didn’t actually need to change it, so that’s my bad for operating off of an assumption.

I’ll get on making a unit test ASAP

smoogipoo commented:
!diffcalc

smoogipoo commented:
!diffcalc

Finadoggie committed:
Revert "use LargeTickHit instead of LargeTickMiss"

This reverts commit 1337b7eb41d8a3e225bf7e8e1d0b9138dca88927.

Finadoggie commented:
the commit above that I reverted was made off of an incorrect assumption of mine 

I saw that HitResult.LargeTickMiss was not returned in api calls and assumed that that meant I couldn't use it in pp calc (made the commit), but was then informed that it can still be used as it should be calculated locally by any pp calc tools and that using said data is still fine (hence the reversion)

Finadoggie committed:
add large tick misses back into effectivemisscount

Finadoggie commented:
After a discussion in the pp server, there was a (mostly) unanimous agreement that large tick misses should count as misses since they break combo, as there would be a large discrepancy between stable and lazer scores otherwise. However, they are capped to effectiveMissCount since they could otherwise lead to absurdly punishing misscounts. This may be addressed in a future rework, but is fine for right now.

Finadoggie committed:
merge effectivemisscount functions
having two functions was unnecessary

Finadoggie committed:
Addressed code quality concerns

Finadoggie committed:
Only clamp estimated miss count with relevant statistics

peppy commented:
!diffcalc

Finadoggie commented:
Updated the original post to hopefully make the goals of the PR more clear

Finadoggie commented:
@smoogipoo multiple people have told me that a unit test is no longer required since I've cleaned up and simplified everything, do you still want me to write one?

minisbett commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -46,6 +48,12 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
             countMiss = score.Statistics.GetValueOrDefault(HitResult.Miss);
             effectiveMissCount = calculateEffectiveMissCount(osuAttributes);
 
+            if (!usingClassicSliderAccuracy)
+            {
+                countSliderEndsDropped = osuAttributes.SliderCount - score.Statistics.GetValueOrDefault(HitResult.SliderTailHit);
+                countLargeTickMiss = score.Statistics.GetValueOrDefault(HitResult.LargeTickMiss);
+            }
+
```
```suggestion
            if (!usingClassicSliderAccuracy)
            {
                countSliderEndsDropped = osuAttributes.SliderCount - score.Statistics.GetValueOrDefault(HitResult.SliderTailHit);
                countLargeTickMiss = score.Statistics.GetValueOrDefault(HitResult.LargeTickMiss);
            }

            effectiveMissCount = calculateEffectiveMissCount(osuAttributes);
```
Effective misscount should not be calculated before that info is available

Finadoggie replied:
fixed, thanks!

Finadoggie committed:
Fix variables being used before being assigned
slightly miffed by the lack of build errors but oh well

smoogipoo commented:
This is honestly harder to read now than before, now that it has 4 separate places dealing with "usingClassicSliderAccuracy`.

stanriders committed:
Refactor and add comments

stanriders committed:
Fix effectiveMissCount being calculated wrong

minisbett commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -44,7 +57,36 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+                    effectiveMissCount = Math.Min(effectiveMissCount, totalImperfectHits);
+                }
+                else
+                {
+                    double fullComboThreshold = attributes.MaxCombo - countSliderEndsDropped;
```
```suggestion
                {
                    // Consider that full combo is maximum combo minus dropped slider tails since they don't contribute to combo but also don't break it
                    double fullComboThreshold = attributes.MaxCombo - countSliderEndsDropped;
```
A comment to stay consistent with the other if-branch would be nice

stanriders replied:
Repeating the same comment is redundant

minisbett replied:
fair just wanted to mention it incase it is desired

minisbett commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -25,6 +25,19 @@ public class OsuPerformanceCalculator : PerformanceCalculator
+        /// <summary>
+        /// Missed slider ticks that includes missed reverse arrows. Will only be correct on non-classic scores
+        /// </summary>
+        private int countSliderTickMiss;
```
I feel like renaming this to `countLargeTickMiss` makes it alot clearer what it is because despite the comment reading it anywhere else in code could put up confusion... I don't see why it shouldn't be named 1:1 after the hit result it represents.

stanriders replied:
"LargeTick" does not represent anything on its own unless you're aware of what it actually contains

minisbett replied:
I feel like it's still a more meaningful term since it matches the hit result and also the "Large Tick Misses" field I added to osu-tools GUI and soon to CLI.

minisbett commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -25,6 +25,19 @@ public class OsuPerformanceCalculator : PerformanceCalculator
+        /// <summary>
+        /// Amount of missed slider tails that don't break combo. Will only be correct on non-classic scores
+        /// </summary>
+        private int countSliderEndsDropped;
```
Since slider-related code uses the term "tail" (and the hit result does too), I'd suggest renaming this to `countSliderTailMiss` or `countSliderTailsDropped`. I'd prefer to consider "miss" the opposite of hit even for slider tails but calling them "dropped" is arguable

minisbett commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -124,8 +166,22 @@ private double computeAimValue(ScoreInfo score, OsuDifficultyAttributes attribut
             {
-                double estimateSliderEndsDropped = Math.Clamp(Math.Min(countOk + countMeh + countMiss, attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders);
-                double sliderNerfFactor = (1 - attributes.SliderFactor) * Math.Pow(1 - estimateSliderEndsDropped / estimateDifficultSliders, 3) + attributes.SliderFactor;
+                double estimateImproperlyFollowedDifficultSliders;
+
+                if (usingClassicSliderAccuracy)
+                {
+                    // When the score is considered classic (regardless if it was made on old client or not) we consider all missing combo to be dropped difficult sliders
+                    int maximumPossibleDroppedSliders = totalImperfectHits;
+                    estimateImproperlyFollowedDifficultSliders = Math.Clamp(Math.Min(maximumPossibleDroppedSliders, attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders);
+                }
+                else
+                {
+                    // We add tick misses here since they too mean that the player didn't follow the slider properly
+                    // We however aren't adding misses here because missing slider heads has a harsh penalty by itself and doesn't mean that the rest of the slider wasn't followed properly
+                    estimateImproperlyFollowedDifficultSliders = Math.Min(countSliderEndsDropped + countSliderTickMiss, estimateDifficultSliders);
+                }
+
+                double sliderNerfFactor = (1 - attributes.SliderFactor) * Math.Pow(1 - estimateImproperlyFollowedDifficultSliders / estimateDifficultSliders, 3) + attributes.SliderFactor;
```
```suggestion
                double improperlyFollowedDifficultSliders;

                if (usingClassicSliderAccuracy)
                {
                    // When the score is considered classic (regardless if it was made on old client or not) we consider all missing combo to be dropped difficult sliders
                    int maximumPossibleDroppedSliders = totalImperfectHits;
                    improperlyFollowedDifficultSliders = Math.Clamp(Math.Min(maximumPossibleDroppedSliders, attributes.MaxCombo - scoreMaxCombo), 0, estimateDifficultSliders);
                }
                else
                {
                    // We add tick misses here since they too mean that the player didn't follow the slider properly
                    // We however aren't adding misses here because missing slider heads has a harsh penalty by itself and doesn't mean that the rest of the slider wasn't followed properly
                    improperlyFollowedDifficultSliders = Math.Min(countSliderEndsDropped + countSliderTickMiss, estimateDifficultSliders);
                }

                double sliderNerfFactor = (1 - attributes.SliderFactor) * Math.Pow(1 - improperlyFollowedDifficultSliders / estimateDifficultSliders, 3) + attributes.SliderFactor;
```
The variable name is quite long and since it's not always estimated so it doesn't necessarily have a valuable meaning besides making the code more straining to read.

stanriders replied:
I prefer longer variable names that encapsulates what it actually is instead of hiding the fact that it can be estimated (and will be for billions of scores)

minisbett replied:
hmm idk, I feel like there isn't much difference between "estimated" when it can or cannot be vs. not saying it when it can or cannot be

tsunyoku commented:
we'll need a new sheet following recent commits, https://github.com/ppy/osu/pull/27691/commits/98800fea71e35552cb0fa8080c282e78bb4723a3 will have resulted in value changes even before StanR's refactors

smoogipoo commented:
Am going to use this opportunity to also test the new PP processing command.

!diffcalc
SCORE_PROCESSOR_A=https://github.com/ppy/osu-queue-score-statistics/pull/284
SCORE_PROCESSOR_B=https://github.com/ppy/osu-queue-score-statistics/pull/284

smoogipoo commented:
!diffcalc

smoogipoo commented:
@ppy/osu-pp-committee please check above sheet!

stanriders commented:
@smoogipoo all good!

apollo-dw reviewed (approved):
one more for luck :]

smoogipoo reviewed (approved):
Reads fine to me now
